src/main/groovy/org/savantbuild/plugin/java/testng/JavaTestNGPlugin.groovy

```diff
@@ -363,4 +375,49 @@ class JavaTestNGPlugin extends BaseGroovyPlugin {
     }
   }
 
+  def produceCodeCoverageReports() {
+    def loader = new ExecFileLoader()
+    // produced by the Java Agent we instrumented our test run with
+    def execFile = getCodeCoverageFile()
+    if (!execFile.exists()) {
+      fail("${execFile} was not found")
+    }
+    loader.load(execFile)
+    def builder = new CoverageBuilder()
+
+    def analyzer = new Analyzer(loader.executionDataStore, builder)
+    // refine our analysis to only include the main JAR that we publish
+    def coverageClassPath = dependencyPlugin.classpath {
+      project.publications.group("main").each { publication -> path(location: publication.file.toAbsolutePath()) }
+    }
+    // add each JAR from the Savant main publication group to be analyzed
+    coverageClassPath.paths.each { p ->
+      analyzer.analyzeAll(p.toFile())
+    }
+
+    def bundle = builder.getBundle("JaCoCo Coverage Report - ${project.name}")
+    def formatter = new HTMLFormatter()
+    def reportDirectory = new File(project.directory.toFile(), "build/coverage-reports")
+    def visitor = formatter.createVisitor(new FileMultiReportOutput(reportDirectory))
+    visitor.visitInfo(loader.sessionInfoStore.infos,
+        loader.executionDataStore.contents)
+    def sourceLocator = new MultiSourceFileLocator(4)
+    sourceLocator.add(new DirectorySourceFileLocator(new File(project.directory.toFile(), "src/main/java"),
+        "utf-8",
+        4))
+    visitor.visitBundle(bundle, sourceLocator)
+    visitor.visitEnd()
+  }
+
+  File getCodeCoverageFile() {
+    project.directory.resolve("build/jacoco.exec").toAbsolutePath().toFile()
+  }
+
+  String getCodeCoverageArguments() {
+    if (!settings.codeCoverage) {
+      return ""
+    }
+    def jacocoPath = AgentJar.extractToTempLocation()
+    " -javaagent:${jacocoPath}=destfile=${getCodeCoverageFile()}"
```

Comment:
There are additional arguments you can supply (see https://www.jacoco.org/jacoco/trunk/doc/agent.html) but "Except for performance optimization or technical corner cases this option is normally not required" so leaving this alone